src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -380,6 +380,19 @@ class AuthenticationCompleteException : Exception { }
                 return matchingEntry.Password;
             }
 
+            if (ProvidePasswordCallback != null)
+            {
+                Log.Trace("Taking password from GetDynamicPassword delegate");
+                try
+                {
+                    return ProvidePasswordCallback(Host, Port, Settings.Database!, username);
```

Comment:
Database is a bit similar to username, in that we need to pass it down rather than access Settings directly (the fact that you had to put a bang is a hint). [In NpgsqlConnector.Open()](https://github.com/npgsql/npgsql/blob/ba045922f2419a18c4c8bd61c8a22b63e13843e3/src/Npgsql/NpgsqlConnector.cs#L385) we default the database to the username if not specified, and this also needs to flow down.

Reply:
yeah i spotted this
https://github.com/npgsql/npgsql/blob/ca75272f55730052da721b99bed76f1df237013c/src/Npgsql/NpgsqlConnector.cs#L382-L383

Reply:
so settings database should be safe to use - hence the bang

Reply:
My bad - I misread the code. Not sure I like it that way, but that's another question - your PR is fine the way it is.